feat(bootstrap): implement bootstrap status facade method

SimoneDutto opened the pull request:
## Description

### First commit

Implement bootstrap status facade method by creating the bootstrap manager and exposing API to get status and logs of a bootstrap job.

To do so I needed to add:
- db method to GetJob
- jobID in the context for the `job` func
- tests

```terminal
[10:15:34] ➜  jimm git:(bootstrap-show-facade) ✗ stress -count 100 ./bootstrap.test -check.f bootstrapManagerSuite.TestGetBootstrapStatusAndLogs
5s: 48 runs so far, 0 failures, 12 active
9s: 100 runs total, 0 failures
```

### Second commit
Before I wasn't able to stress test TestGetBootstrapStatusAndLogs because during SetupTestOFGAClient
in RemoveStore it does "DELETE FROM authorization_model WHERE store = (SELECT id FROM store WHERE
name = %s"). This returns an error if the subquery returns multiple entries. And this can happen if
we stress test a test because it is setting up the store multiple times with the same name.

SimoneDutto committed:
feat(bootstrap): implement bootstrap status facade method

Implement bootstrap status facade method by creating the bootstrap manager and exposing API to get
status and logs of a bootstrap job.

SimoneDutto commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+	c.Assert(err, qt.IsNil)
+}
+
+func (s *bootstrapManagerSuite) TestGetBootstrapStatusAndLogs(c *qt.C) {
```
very proud of this test

SimoneDutto commented on internal/jimm/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,65 @@
+// GetBootstrapStatusAndLogs retrieves the status and logs of a bootstrap job.
+// It requires the user to be an admin and returns the status, error message, logs,
+// and a watermark for pagination.
+func (b *bootstrapManager) GetBootstrapStatusAndLogs(ctx context.Context, _ *openfga.User, jobId uuid.UUID, offset int) (params.BootstrapStatusResponse, error) {
```
here i have a discussion point:
- we check the user a jimm admin in `jujuapi` pkg
- should we check it again here?
- trust the caller has already checked the user is a jimm admin?

ale8k replied:
Trust it, it's ok.

kian99 replied:
Imo it's sufficient as is with the check in `jujuapi`.

kian99 commented on internal/jimm/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+type bootstrapManager struct {
+	jobtracker JobTracker
+	store      *db.Database
```
Why is there a job tracker and the store? Is that because the logs need the DB but you want to test with mock jobs?

SimoneDutto replied:
because the job tracker is just to start, stop, get status for the job.
However to get the bootstrap logs you can't use the jobtracker but you need the database

ale8k commented on internal/jimm/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+// JobTracker interface defines the methods required for job tracking.
+type JobTracker interface {
```
s/JobTracker/JobTrackerStore

SimoneDutto replied:
why? it's not really a store, since it runs a go-routine for you

kian99 commented on internal/jimm/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,65 @@
+	if authSvc == nil {
+		return nil, errors.E("group authorisation service cannot be nil")
+	}
+	return &bootstrapManager{
```
We can check the jobtracker is not nil too.

ale8k commented on internal/jimm/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,65 @@
+
+type bootstrapManager struct {
+	jobtracker JobTracker
+	store      *db.Database
```
We shouldn't need the entire db, be good to define an interface here so we can mock it.

kian99 commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+	)
+	c.Assert(err, qt.IsNil)
+	watermark := 0
+	for i := 0; i < numLogs/batchSize+1; i++ {
```
Replace i with `batch`?

kian99 commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+		response, err := s.manager.GetBootstrapStatusAndLogs(ctx, s.adminUser, jobId, watermark)
+		c.Assert(err, qt.IsNil)
+		logs := []string{}
+		for j := 0; j < int(math.Min(float64(batchSize), float64(numLogs-i*batchSize))); j++ {
```
Replace `j` with `logNumber`?

ale8k replied:
This kinda reads like a cluster of code, please save my feelings 😝

ale8k commented on internal/jimm/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,65 @@
+		return params.BootstrapStatusResponse{}, errors.E(op, "failed to get job status", err)
+	}
+
+	logs, newOffset, err := b.store.QueryBootstrapLog(ctx, jobId, offset)
```
Can we call it loggies? (Internal joke I did in the query haha) 😜

SimoneDutto replied:
for sure!

ale8k commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+func (s *bootstrapManagerSuite) TestGetBootstrapStatusAndLogs(c *qt.C) {
+	ctx := c.Context()
+	read := make(chan struct{})
+	defer close(read)
```
Some space between some vars would help readability for me like uhm:
```go
	ctx := c.Context()

	read := make(chan struct{})
    write := make(chan struct{})
	defer close(read)
	defer close(write)

	numLogs := 101
	batchSize := 10
```

But feel free to ignore, and hurt my feelings 😂

ale8k commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+	defer close(write)
+	numLogs := 101
+	batchSize := 10
+	jobId, err := s.jobTracker.Run(ctx,
```
And like here, we got ctx on 62, but rest of params on new lines. Again hurt my feelings if you want haha.

ale8k commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+				err := s.db.AddBootstrapLog(ctx, jobId, "bootstrap logs "+fmt.Sprint(rune(i)))
+				c.Check(err, qt.IsNil)
+			}
+			write <- struct{}{}
```
Why do we need this final write and read?

SimoneDutto replied:
left a comment

ale8k commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+			<-read
+			return nil
+		},
+		1*time.Minute,
```
I'd probably set this quite a bit shorter as it is your only way to clean up the routine. And additionally, in your job routine we can check the test context isn't done, and if it is we return. Just as a safety precaution to ensure the routine is cleaned up.

SimoneDutto replied:
isn't always cleanup just by the fact we return from the `job` func?

SimoneDutto committed:
test(name with uuid): add uuid to test's db name and openfga db name

Before I wasn't able to stress test TestGetBootstrapStatusAndLogs because during SetupTestOFGAClient
in RemoveStore it does "DELETE FROM authorization_model WHERE store = (SELECT id FROM store WHERE
name = %s"). This returns an error if the subquery returns multiple entries. And this can happen if
we stress test a test because it is setting up the store multiple times with the same name.

kian99 commented on internal/jobtracker/jobtracker_test.go:
```diff
@@ -105,6 +105,21 @@ func (s *jobTrackerSuite) TestRun_JobSetSuccessful(c *qt.C) {
+	testCtx := c.Context()
+
+	aJobWithIdInContext := func(ctx context.Context) error {
+		_, ok := jobtracker.JobIdFromContext(ctx)
```
Minor but I suggest to check that the value is non-zero or something along those lines.

ale8k commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+	write := make(chan struct{})
+	defer close(write)
+	numLogs := 101
+	batchSize := 10
```
I think a simpler test where we just read 1, then 2, then 1 again would read better than batching.

ale8k commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+		for j := 0; j < int(math.Min(float64(batchSize), float64(numLogs-i*batchSize))); j++ {
+			logs = append(logs, "bootstrap logs "+fmt.Sprint(rune(i*batchSize+j)))
+		}
+		c.Check(response.Logs, qt.DeepEquals, logs)
```
We can assert this?

SimoneDutto replied:
Nope, because we are in a go routine

ale8k commented on internal/jimm/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,137 @@
+
+	// check last batch is empty.
+	response, err := s.manager.GetBootstrapStatusAndLogs(ctx, s.adminUser, jobId, watermark)
+	c.Assert(response.Status == string(dbmodel.StatusSuccessful) || response.Status == string(dbmodel.StatusRunning), qt.IsTrue)
```
Shouldn't we only wait for it to be successful? Can pinch my polling code from the job tracker test

SimoneDutto replied:
I dind't want to poll, so i'm just checking the job is either running or successful.
I think it's fine because here we are not testing that the jobtracker is marking the job as successful

ale8k commented on internal/jujuapi/jimm.go:
```diff
@@ -558,3 +561,19 @@ func (r *controllerRoot) PrepareModelMigration(ctx context.Context, args apipara
 	return resp, nil
 }
+
+// BootstrapStatus retrieves the status of a bootstrap job and its logs.
```
It also returns the next watermark for consumers to use.

ale8k commented on internal/jujuapi/jimm.go:
```diff
@@ -558,3 +561,19 @@ func (r *controllerRoot) PrepareModelMigration(ctx context.Context, args apipara
+func (r *controllerRoot) BootstrapStatus(ctx context.Context, req apiparams.BootstrapStatusRequest) (apiparams.BootstrapStatusResponse, error) {
+	const op = errors.Op("jujuapi.BootstrapStatus")
+
+	if !r.user.JimmAdmin {
```
Now, thinking about this, admin is ok for now, but maybe we should be checking they own the credential too. Admins should never be able to use others credentials right? Or users in general, so that means someone (who might not be the admin) started a bootstrap but the admin can view their logs? Maybe a little discussion point for later as I'm not sure we'd doing permissions right here. I think we should be checking admin + owns credential for this bootstrap (idk how we'd check that here though 😢)

SimoneDutto replied:
that's the reason why the bootstrap manager has the openfga client and i'm passing the user down to the methods.
So we can later implement some fine-grained checks

ale8k commented on internal/jujuapi/jimm_unit_test.go:
```diff
@@ -80,3 +105,48 @@ func (s *jimmSuite) TestPrepareModelMigration_InvalidUserMapping(c *gc.C) {
+	c.Assert(errors.ErrorCode(err), gc.Equals, errors.CodeNotFound)
+
+	// Test unauthorized user
+	root = newTestControllerRoot(jimm, "alice@canonical.com", false)
```
Yeah as I read this it is feeling like it doesn't make sense that an admin can view the progress of a bootstrap that might not be theirs.

ale8k commented on pkg/api/params/params.go:
```diff
@@ -560,3 +560,23 @@ type PrepareModelMigrationResponse struct {
+// BootstrapStatusRequest holds the request to get the status of a bootstrap job.
+type BootstrapStatusRequest struct {
+	// JobID is the ID of the bootstrap job to get the status for.
+	JobID string `json:"job-id"`
```
Missing yaml tags

ale8k commented on pkg/api/params/params.go:
```diff
@@ -560,3 +560,23 @@ type PrepareModelMigrationResponse struct {
+	// JobID is the ID of the bootstrap job to get the status for.
+	JobID string `json:"job-id"`
+	// Watermark is the line number to start reading logs from.
+	Watermark int `json:"watermark"`
```
This is omittable no? 0 is fine?

SimoneDutto replied:
we would always set this field from the bootstrap manager, so it would never be omitted.

ale8k reviewed (approved):
Beeeautiful. Thanks for this! Some minor comments (plz do them) but LGTM otherwise!

SimoneDutto committed:
chore(pr comments): pr comments

SimoneDutto committed:
Merge branch 'v3' into bootstrap-show-facade
